### DISCUSSION TITLE
Issue with repeater (nested non associatve array)

### DISCUSSION THREAD

**Discussion by underdpt** [filamentphp/filament#6985] [Help]
Hello,

When trying to use a `Repeater` to save a non associative array that's inside a json column, I'm getting an error when loading the model edit page after saving:

```
Filament\Forms\ComponentContainer::getRawState(): Return value must be of type Illuminate\Contracts\Support\Arrayable|array, string returned
```

This is my repeater:

```php
Forms\Components\Repeater::make('personalizables.1.opcion')
    ->schema([
        Forms\Components\TextInput::make('')
            ->maxLength(128),
    ])
    ->label('Opciones para el campo libre 1')
    ->defaultItems(3)
    ->minItems(1)
    ->maxItems(10)
    ->disableItemMovement()
    ->itemLabel('Opción')
    ->createItemButtonLabel('Añadir nueva opción')
    ->columnSpanFull(),
```

As you can see, it's nested on a `personalizables` column which has at most two fixed items, the repeater is saved on `personalizables.1.opcion` as an array.  This is the json that's saved on the database:

```json
{
    "1": {
        "activo": true,
        "nombre": "Nombre guay",
        "opcion": [
            "myopcion"
        ],
        "guardado": true,
        "usar_en_informes": false,
        "mostrar_denunciante": false,
        "obligatorio_denunciante": false
    },
    "2": {
        "opcion": [],
        "guardado": true
    }
}
```

The saved data it's correct.

It's there something I'm doing wrong?

Thanks!

**Comment by pboivin** [comment]
Can you save the repeater data normally and instead, transform it when you need it out of the database?

**Reply by danharrin** [reply]
The problem is because the repeater data in the structure is an array of strings, not an array of arrays as it should be

**Reply by underdpt** [reply]
Unfortunately this is an app I'm upgrading to Filament, and this data is already populated on some records, so I have to maintain this structure.

I've been doing some tests, and this change make the trick:

`\Filament\Forms\Concerns\HasState::getRawState()` (line 187):

From:

```php
    public function getRawState(): array | Arrayable
    {
        return data_get($this->getLivewire(), $this->getStatePath()) ?? [];
    }
```

To:
```php
    public function getRawState(): array | Arrayable
    {
        $rawState = data_get($this->getLivewire(), $this->getStatePath()) ?? [];

        return is_string($rawState) ? [$rawState] : $rawState;
    }
```

It has no effect on repeaters working as intended, and would allow to use repeaters to save a collection of strings, that might be usefult to others.

Would you accept a PR with this change?

If the answer is negative, is there a way for me to overload this function with a custom component?

**Reply by sawirricardo** [reply]
or simply transform the existing structure into the structure the repeater use?

**Reply by underdpt** [reply]
Yes, I've considered that, but that would be a huge work as we would need to change also all the places where that data is used, or other methods that also update that structure. I would like to avoid that, not only for the work but because is easy to introduce new bugs.

**Reply by sawirricardo** [reply]
You can try using [`dehydrateStateUsing()`](https://filamentphp.com/docs/2.x/forms/advanced#dehydration) if I'm not mistaken to transform the data (`afterStateHydrated()` may also work too)

Another idea is to use the eloquent's accessor and mutator, so you can implement the logic you wrote at `getRawState()`

**Reply by bartlomiej-przymus** [reply]
I had the same issue and I have solved it by doing this in filament v3.
```php
Repeater::make('options')
  ->simple(
    TextInput::make('')
    ->required()
    ->maxLength(255)
    ->string()
  )
  ->afterStateHydrated(
    function (?array $state, string $operation) {
      $state = ($operation === 'create') ? [] : $state;
    })
  ->mutateDehydratedStateUsing(
    fn ($state) => $state = array_values($state)
  )
  ->reorderable(false),
```
When hydrating data in new form it was filling state of new TextInput component with message [object Object] and some deeply nested array hence why I was forced to clean it with overriding `afterStateHydrated()` method.
On save we need to get rid of all uuid's that are generated when Repeater creates new instances of TextInput component
by using `mutateDehydratedStateUsing()`

Issues that remain:
It seems reordering does not work if we override data this way but I didn't need it anyway. There are methods to override reordering actions too so it might be solvable.
